sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -56,33 +58,93 @@ case class CreateArray(children: Seq[Expression]) extends Expression {
   }
 
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
-    val arrayClass = classOf[GenericArrayData].getName
-    val values = ctx.freshName("values")
-    ctx.addMutableState("Object[]", values, s"this.$values = null;")
+    val array = ctx.freshName("array")
 
-    ev.copy(code = s"""
-      this.$values = new Object[${children.size}];""" +
+    val et = dataType.elementType
+    val evals = children.map(e => e.genCode(ctx))
+    val isPrimitiveArray = ctx.isPrimitiveType(et)
+    val primitiveTypeName = if (isPrimitiveArray) ctx.primitiveTypeName(et) else ""
+    val (preprocess, arrayData, arrayWriter) =
+      GenArrayData.getCodeArrayData(ctx, et, children.size, isPrimitiveArray, array)
+
+    ev.copy(code =
+      preprocess +
       ctx.splitExpressions(
         ctx.INPUT_ROW,
-        children.zipWithIndex.map { case (e, i) =>
-          val eval = e.genCode(ctx)
-          eval.code + s"""
-            if (${eval.isNull}) {
-              $values[$i] = null;
+        evals.zipWithIndex.map { case (eval, i) =>
+          eval.code +
+            (if (isPrimitiveArray) {
+              (if (!children(i).nullable) {
+                s"\n$arrayWriter.write($i, ${eval.value});"
+              } else {
+                s"""
+                if (${eval.isNull}) {
```

Comment:
What do you mean "manually optimize"? I naively generate the code without optimization. While I could optimize the code by eliminating the generation of `$arrayWriter.setNull$primitiveTypeName($i);` if `${eval.isNull} == "false"`, I did not apply it here.